include/pybind11/pybind11.h

```diff
@@ -1100,7 +1100,9 @@ class class_ : public detail::generic_type {
     template <typename T>
     static void init_holder_helper(instance_type *inst, const holder_type * /* unused */, const std::enable_shared_from_this<T> * /* dummy */) {
         try {
-            new (&inst->holder) holder_type(std::static_pointer_cast<typename holder_type::element_type>(inst->value->shared_from_this()));
+            new (&inst->holder) holder_type(std::static_pointer_cast<typename holder_type::element_type>(
+                static_cast<T*>((void*)inst->value)->shared_from_this()
```

Comment:
If `T` is a base, a single `static_cast` should be enough -- no need to go through `void*` (which is essentially reinterpret casting).

Alternatively, the casting can be avoided altogether by using the currently commented out parameter:
```c++
// existing unnamed parameter
const std::enable_shared_from_this<T> * /* dummy */

// -> name it
const std::enable_shared_from_this<T> *esft

// -> use it
esft->shared_from_this()
```

And as @jagerman has already mentioned, a minimal test case would be very good here.